### ISSUE TITLE
Auto insert custom title on weekly, monthly, and yearly calendars?

### ISSUE THREAD

**Issue by anthrolisp** [bastibe/org-journal#153]
I just saw "Daily weekly monthly yearly journal implementation" #143

Is there anyway to have something like "#+title: MONTH YEAR" or "#+title: Week No. YEAR" automatically inserted upon new file creation?

Many thanks in advance!

**Comment by cslux** [comment]
Not at the moment. I will add something.

**Comment by anthrolisp** [comment]
Thank you!

Somewhat related question: with  setq org-journal-file-type set to weekly, is it possible to have org-journal create all 7 days of the week when creating the file? I like to plan my week in advance, and being able to see the whole week and move items around in the file would be greatly appreciated. Thanks again!

**Comment by bastibe** [comment]
You should be able to create future entries through the calendar, or `org-journal-new-entry`.

**Comment by cslux** [comment]
> is it possible to have org-journal create all 7 days of the week when creating the file?

Currently, it is only possible to insert one future journal entry at a time. But somehow I like that idea.

**Comment by cslux** [comment]
@bastibe Should we add a new defcustom holding a function (user needs to write it) returning a string to be inserted at the top of **new** journal file. This would also solve #94

**Comment by bastibe** [comment]
We already have an `org-journal-after-entry-create-hook`, and `org-journal-date-format` can already be a function. I suppose we could add an `org-journal-new-file-hook` or something like that. However, this function would most likely contain some pretty complicated code, as it would probably need to extract the file's date, and behave differently for daily, weekly, or monthly files. And if it needs to be complex already, one might as well add that one conditional, and use `org-journal-after-entry-create-hook` instead, but trigger only on recently-created files.

In summary, I would tentatively reject adding a new hook for this, on the grounds that there is no easy way of using that hook that wouldn't be similarly easy with the existing hook. Maybe a section in the README that explains how to do this with `org-journal-after-entry-create-hook` would be a more appropriate solution.

**Comment by snuffop** [comment]
@cslux   I have used a bit of Howard Abrams code  https://github.com/howardabrams/dot-files  to use the emacs template system to add the headers to each new journal file.

`(defun org-rename-header (label)
  "Rename the current section's header to LABEL, and moves the
point to the end of the line."
  (interactive (list
                (read-string "Header: "
                             (substring-no-properties (org-get-heading t t t t)))))
  (org-back-to-heading)
  (replace-string (org-get-heading t t t t) label))
`

**Comment by anthrolisp** [comment]
It's also minimal work to manually enter a title for 1 (yearly), 12 (monthly), or even 52 (weekly) files a year.

**Comment by cslux** [comment]
Using `org-journal-date-format` is a good idea.

I will add this to the readme. Here is a working example.

``` emacs-lisp
(defun org-journal-date-format-func (time)
  "Custom function to insert journal date header, 
and some custom text on a newly created journal file."
  (when (= (buffer-size) 0)
    (insert
     (pcase org-journal-file-type
       (`daily "#+TITLE: Daily Journal")
       (`weekly "#+TITLE: Weekly Journal")
       (`monthly "#+TITLE: Monthly Journal")
       (`yearly "#+TITLE: Yearly Journal"))))
  (concat org-journal-date-prefix (format-time-string "%A, %x" time)))

(setq org-journal-date-format 'org-journal-date-format-func)
```

**Comment by sharad-shrestha** [comment]
I used the `org-journal-date-format-func` suggested above; that is,

```
(defun org-journal-date-format-func (time)
  "Custom function to insert journal date header, 
and some custom text on a newly created journal file."
  (when (= (buffer-size) 0)
    (insert
     (pcase org-journal-file-type
       (`daily "#+TITLE: Daily Journal")
       (`weekly "#+TITLE: Weekly Journal")
       (`monthly "#+TITLE: Monthly Journal")
       (`yearly "#+TITLE: Yearly Journal"))))
  (concat org-journal-date-prefix (format-time-string "%A, %x" time)))

(setq org-journal-date-format 'org-journal-date-format-func)
```

When I set `(org-journal-file-type 'weekly)`, I get correct entries in the journal file, which is:

```
#+TITLE: Weekly Journal

* Tuesday, 2019-06-25
:PROPERTIES:
:CREATED:  20190625
:END:
** 19:33
```

I also get correct entries in the journal file when I set `org-journal-file-type` to `'monthly`, or `'yearly`.

However, when I set `  (org-journal-file-type 'daily)`, I get the following incorrect entry in the journal file:
```
* Tuesday, 2019-06-25#+TITLE: Daily Journal

** 19:40
```

To get around this error, in `org-journal.el`, at line 472, I added 4 lines of code to the function `org-journal-new-entry`, as shown below:
```
      ;; Create new journal entry if there isn't one.
      (let ((entry-header
             (if (functionp org-journal-date-format)
                 (funcall org-journal-date-format time)
               (concat org-journal-date-prefix
                       (format-time-string org-journal-date-format time)))))
        (goto-char (point-min))
        (unless (search-forward entry-header nil t)
          ;; Insure we insert the new journal header at the correct location
          ;; ======= I ADDED THE FOLLOWING 4 LINES ==========
          (if (org-journal-daily-p)
              (goto-char (point-max))
              (forward-line)
            )
          ;; ========================================
          (unless (org-journal-daily-p)
            (let ((date (decode-time time))
                  (dates (sort (org-journal-file->calendar-dates (buffer-file-name))
                               (lambda (a b)
                                 (calendar-date-compare (list b) (list a)))))
                  found)
```

With this addition, I get the correct entry in the journal file, which is:

```
#+TITLE: Daily Journal

* Tuesday, 2019-06-25
** 19:51 (07:51 PM)
```

**Comment by cslux** [comment]
@sharad-shrestha Thank you! Should be fixed with 2135c5f.

**Comment by sharad-shrestha** [comment]
Thank you @cslux .

Now, when my journal directory does not contain any files, insertion of daily custom header works fine, with the output file `20190626` containing:
```
#+TITLE: Daily Journal

* Wednesday, 2019-06-26
** 10:48 (10:48 AM)
```

If my journal directory already contains file `20190625` with the following contents:
```
#+TITLE: Daily Journal

* Tuesday, 2019-06-25
** 10:48 (10:48 AM)
```
and I run `C-c C-j`, it creates a new daily journal file `20190626` with correct contents:
```
#+TITLE: Daily Journal

* Wednesday, 2019-06-26
** 10:48 (10:48 AM)
```
However, then the program proceeds to search for `TODO` headers on the previous daily journal file, at which point it generates error with the following error messages:
```
Debugger entered--Lisp error: (error "Before first headline at position 1 in buffer 20190625")
  signal(error ("Before first headline at position 1 in buffer 20190625"))
  error("Before first headline at position %d in buffer %s" 1 #<buffer 20190625>)
  (condition-case nil (outline-back-to-heading invisible-ok) (error (error "Before first headline at position %d in buffer %s" (point) (current-buffer))))
  org-back-to-heading()
...
...
```
If I remove the title header from the previous day's journal file `20100625`, which now looks like the following:
```
* Tuesday, 2019-06-25
** 10:48 (10:48 AM)
```
and I run `C-c C-j`, it creates new daily journal file `20190626` with correct contents:
```
#+TITLE: Daily Journal

* Wednesday, 2019-06-26
** 10:48 (10:48 AM)
```
So, it looks like the program is expecting the first line in the previous day's journal file to contain the header line of the form `* Tuesday, 2019-06-26`.

**Comment by cslux** [comment]
@sharad-shrestha Should be fixed with the commit c522039.

Sorry :-) 

Thanks, Chris!

**Comment by sharad-shrestha** [comment]
Thank you, @cslux .

The custom title for daily, weekly, monthly, and yearly `org-journal-file-type` all work correctly now when I execute `C-c C-j` command.

When I execute `C-u C-c C-j`, I observed the following behavior for `daily`, `weekly`, `monthly`, and `yearly` `org-journal-file-type`.  If the journal file for the day (or week, month, or year) does not exist, it creates a new file with the following entry (shown below for daily case):
```
#+TITLE: Daily Journal

* Wednesday, 2019-06-26
```
It behaves similarly for `weekly`, `monthly`, and `yearly` `org-journal-file-type`.

**I think, may be a better response is for the program to simply generate a message (in the echo area) informing that the file does not exist (and not create a new file).** 

If the journal file for the day (for `org-journal-file-type`=`'daily`) already exists and contains the following entries:
```
#+TITLE: Daily Journal

* Wednesday, 2019-06-26
** 23:17 (11:17 PM) Some text
```
and if I execute the command `C-u C-c C-j`, it opens up the file without adding any new entry, which is as expected.

If the journal file for the week (for `org-journal-file-type` = `'weekly`) already exists and contains the following entries:
```
#+TITLE: Weekly Journal

* Tessday, 2019-06-25
:PROPERTIES:
:CREATED:  20190625
:END:
** 23:28 (11:28 PM) Some text

```
and if I execute the command `C-u C-c C-j`, it opens up the file and adds today's date as shown below:
```
#+TITLE: Weekly Journal

* Tessday, 2019-06-25
:PROPERTIES:
:CREATED:  20190625
:END:
** 23:28 (11:28 PM) Some text
* Wednesday, 2019-06-26
:PROPERTIES:
:CREATED:  20190626
:END:
```
**I believe it should not have added the entry  `* Wednesday, 2019-06-26`.  It should have simply opened the file without adding any new entry.**

I observed similar program response when the `org-journal-file-type` is `monthly` and `yearly`.

Regards,
Sharad

**Comment by cslux** [comment]
@sharad-shrestha Hey Sharad! Hmm, right! After rereading the function docstring, it shouldn't add a new heading indeed. My fault :-) Good catch!

Thanks,
Chris

**Comment by cslux** [comment]
@sharad-shrestha I updated the docstring. This behaviour is by design.
